pkg/source/unpacker.go

```diff
@@ -0,0 +1,85 @@
+package source
+
+import (
+	"context"
+	"fmt"
+	"io/fs"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Unpacker unpacks source content, either synchronously or asynchronously and
+// returns a Result, which conveys information about the progress of unpacking
+// the source content.
+//
+// If a Source unpacks content asynchronously, it should register one or more
+// watches with a controller to ensure that sources referencing this source
+// can be reconciled as progress updates are available.
+//
+// For asynchronous Sources, multiple calls to Unpack should be made until the
+// returned result includes state StateUnpacked.
+//
+// NOTE: A source is meant to be agnostic to specific source formats and
+// specifications. A source should treat a source root directory as an opaque
+// file tree and delegate source format concerns to source parsers.
+type Unpacker interface {
+	Unpack(context.Context, *Source, client.Object) (*Result, error)
```

Comment:
+1 moving this from `Bundle` to `client.Object`, but it'd be even better if this interface didn't depend on having a kubernetes object. It seems like the primary reason for this needing to exist is to make sure the image unpacker is able to add an owner reference to the pod that it creates.  But it looks like the other source type implementations either don't use it, or use it for what amounts to a unique identifier.

I wonder if there's something we could do to remove this from the interface while also still giving the image unpacker enough information to create a valid owner reference?

Reply:
Maybe the solution is adding an "optional" field to the `Source` type that is of `client.Object` (or maybe even better `metav1.Object`?)? I say "optional" because some sourcing implementations may require it while others may not.